### PR TITLE
Add AWS ARM64 nightly E2E test jobs

### PR DESCRIPTION
**Pull request by LorbusChris**
Blocked on: ~~https://github.com/openshift/ci-tools/pull/2097 and a subsequent `make jobs` on this code here~~ and https://github.com/openshift/release/pull/19729

/cc @jcpowermac 
/cc @deepsm007 
/cc @yselkowitz

### CODE REVIEW

**Comment by deepsm007** [comment]
> Blocked on: [openshift/ci-tools#2097](https://github.com/openshift/ci-tools/pull/2097) and a subsequent `make jobs` on this code here
> 
> /cc @jcpowermac
> /cc @deepsm007
> /cc @yselkowitz

 `make update` instead of `make jobs`

**Comment by jcpowermac** [comment]
Each step that requires the vault secret needs its `-ref.yaml` config mod with:

```
  credentials:
  - namespace: test-credentials
    name: aws-ci-awscred
    mount_path: /var/run/vault/arm/arm-ci-awscred
```

Of course we need to get the creds too ;-)

**Comment by LorbusChris** [comment]
> > Blocked on: [openshift/ci-tools#2097](https://github.com/openshift/ci-tools/pull/2097) and a subsequent `make jobs` on this code here
> > /cc @jcpowermac
> > /cc @deepsm007
> > /cc @yselkowitz
> 
> `make update` instead of `make jobs`, it will fix the config generation issues on the checks below

For either to work the new cluster profile first has to exist: https://github.com/openshift/ci-tools/pull/2097

**Comment by LorbusChris** [comment]
> **The pod could not start because it could not mount the volume "cluster-profile": secret "cluster-secrets-aws-arm64" not found**

Looks like this secret is expected to exist. Might even be possible to stash the awscreds in there instead of referencing them in each step. Will look into that tomorrow.
/cc @jcpowermac

File: core-services/prow/02_config/generate-boskos.py [hunk 1]
```diff
@@ -4,6 +4,13 @@
 
 
 CONFIG = {
+    'aws-arm64-quota-slice': {
+        # Wild guesses.  We'll see when we hit quota issues
+        'us-east-1': 10,
+        'us-east-2': 8,
+        'us-west-1': 8,
+        'us-west-2': 8,
+    },
```
Comment (LorbusChris) [inline core-services/prow/02_config/generate-boskos.py:13]: @coverprice does this seem appropriate to you?

Comment (coverprice) [reply core-services/prow/02_config/generate-boskos.py:13]: No idea. :) Let's wait for the quota bumps, land this, and keep an eye on it.

**Comment by LorbusChris** [comment]
Lifting WIP, hoping this works once quota has been bumped :)

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 2]
```diff
@@ -6,6 +6,21 @@ set -o pipefail
 
 export AWS_SHARED_CREDENTIALS_FILE="${CLUSTER_PROFILE_DIR}/.awscred"
 
+if [[ "${CLUSTER_TYPE}" == "aws-arm64" ]]; then
+  # Hack to avoid importing arm64 release image by using an image override
+  # Find a better way of doing this once https://issues.redhat.com/browse/DPTP-2265 is resolved
+  touch ${SHARED_DIR}/cluster-config.yaml
+  ARM64_RELEASE_IMAGE_INITIAL=registry.ci.openshift.org/ocp-arm64/release-arm64:${BRANCH}
```
Comment (jcpowermac) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:13]: @LorbusChris where is `BRANCH` assigned?
I see libvirt has it created but empty in: 
https://github.com/openshift/release/blob/master/ci-operator/step-registry/ipi/conf/libvirt/ipi-conf-libvirt-ref.yaml#L7

Comment (LorbusChris) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:13]: Oh. I have to check where this is populated there and so the same here

**Comment by LorbusChris** [comment]
/hold wip

`openshift-tests` has to be passed the `--from-repository http://quay.io/multi-arch/community-e2e-images`

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 3]
```diff
@@ -14,19 +23,36 @@ function join_by { local IFS="$1"; shift; echo "$*"; }
...
 if [[ "${SIZE_VARIANT}" == "compact" ]]; then
   workers=0
 fi
 master_type=null
 if [[ "${SIZE_VARIANT}" == "xlarge" ]]; then
-  master_type=m5.8xlarge
+  if [[ "${CLUSTER_TYPE}" == "aws-arm64" ]]; then
+    master_type=m6g.8xlarge
+  else
+    master_type=m5.8xlarge
```
Comment (yselkowitz) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:42]: maybe cleaner to define earlier e.g. `arch_instance_type=m5` or `m6g` depending on `CLUSTER_TYPE`, then just change these definitions to `master_type=${arch_instance_type}.8xlarge` etc.?

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh [hunk 4]
```diff
@@ -14,19 +23,36 @@ function join_by { local IFS="$1"; shift; echo "$*"; }
 
 REGION="${LEASED_RESOURCE}"
 # BootstrapInstanceType gets its value from pkg/types/aws/defaults/platform.go
+architecture="amd64"
 BOOTSTRAP_NODE_TYPE=m5.large
+if [[ "${CLUSTER_TYPE}" == "aws-arm64" ]]; then
+  architecture="arm64"
+  BOOTSTRAP_NODE_TYPE=m6g.large
```
Comment (yselkowitz) [inline ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-commands.sh:30]: see comment about `arch_instance_type`, that would help here too.

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh [hunk 5]
```diff
@@ -37,6 +37,12 @@ function prepare_next_steps() {
 trap 'prepare_next_steps' EXIT TERM
 trap 'CHILDREN=$(jobs -p); if test -n "${CHILDREN}"; then kill ${CHILDREN} && wait; fi' TERM
 
+if [[ "${CLUSTER_TYPE}" == "aws-arm64" ]]; then
+  # Hack to avoid importing arm64 release image by using an image override
+  OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE=registry.ci.openshift.org/ocp-arm64/release-arm64:${BRANCH}
+  export OPENSHIFT_INSTALL_RELEASE_IMAGE_OVERRIDE
+fi
```
Comment (yselkowitz) [inline ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:44]: From where does the `openshift-install` binary used in this file originate, that this override is necessary?

File: ci-operator/config/openshift/release/openshift-release-master__nightly-4.9.yaml [hunk 6]
```diff
@@ -340,6 +340,16 @@ tests:
...
     workflow: openshift-e2e-gcp-realtime
+- as: e2e-aws-arm64
+  interval: 24h
+  steps:
+    cluster_profile: aws-arm64
+    env:
+      BASE_DOMAIN: arm-ci.devcluster.openshift.com
+      BRANCH: "4.9"
+      COMPUTE_NODE_TYPE: m6g.xlarge
+      TEST_ARGS: --from-repository http://quay.io/multi-arch/community-e2e-images
```
Comment (yselkowitz) [inline ci-operator/config/openshift/release/openshift-release-master__nightly-4.9.yaml:351]: The `http://` here needs to be removed.

**Review by yselkowitz — CHANGES_REQUESTED** [review]

**Comment by LorbusChris** [comment]
The e2e-aws-arm64 test now works as expected

/hold cancel

File: ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh (hunk 5 above)
Comment (LorbusChris) [reply ci-operator/step-registry/ipi/install/install/ipi-install-install-commands.sh:48]: To me it looks like the override is always used, but we need a different one from the one that Prow would usually pass.

File: ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-ref.yaml [hunk 7]
```diff
@@ -10,6 +10,10 @@ ref:
       cpu: 10m
       memory: 100Mi
   env:
+  - name: BRANCH
+    default: ""
+    documentation: |-
+      Hack to pass the branch to test. Only consumed by ARM64 tests.
```
[the remark this answers is not in the document]
Comment (LorbusChris) [reply ci-operator/step-registry/ipi/conf/aws/ipi-conf-aws-ref.yaml]: Done

**Comment by yselkowitz** [comment]
+1, test results on ARM are as expected at this moment.

**Review by petr-muller — APPROVED** [review]

**Pull request merged**